Add bonding and vlan tagging support to Bottlerocket

yeazelm opened the pull request:
**Issue number:**
Closes bottlerocket-os/bottlerocket#2369

**Description of changes:**
This will add bonding and vlan tagging support to bottlerocket. There is a lot of code here mostly because we needed to bump the configuration version to 3 and refactor a bunch of logic since there are now different types (kinds) of devices which have different validation logic, testing concerns, etc. This should allow us to add more virtual network devices in the future without so much refactoring. It also moves some of the device-specific code into a new `devices` module for easier understanding of what is a `net_config` concern and what is a "network device" concern. 


**Testing done:**
I've validated this on hardware and qemu. (More testing will come as the PR progresses) The tests all pass as well right now. I also ran clippy to cover those bases as well.



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
[netdog] Add bonding structs for network configuration

This adds the structures and tests that are used to support bonding in
net.toml. This will be enabled via a version 3 of the net_config module
in a future commit. The current implementation only supports mode 1
(active-backup) for bonding and a handful of settings beyond the minimum
set required for bonding. The code is ready for more modes and settings
to be added when the need arises.

Commit:
[netdog] Add vlan structs for network configuration

This adds the stuctures to specify a vlan in net.toml. This allows
deserialization and serialization of vlan structures along with some
tests.

Commit:
[netdog] Add version 3 of network configuration

This adds version 3 of network configuration along with bonding and vlan
support end to end from net.toml to wicked xml. The support now works as
intended for both vlan virtual devices and bonding devices on metal.
This also moves around some of the device files into a new module so
that future expansion of devices is seperated from expansion of network
configuration versions. This also changes the validation logic to add in
traits that allow a generic IP Addressing validation function to run on
any network device that accepts addressing.

stmcginnis commented on PROVISIONING-METAL.md:
```diff
@@ -105,7 +105,30 @@ Please keep in mind that when using static addresses, DNS information must be su
   * `route-metric` (integer): Relative route priority.
 
-Example `net.toml` with comments:
+Version `3` adds in support for bonding and vlan tagging. The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt). Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces. Physical devices like ethernet adapters are considered ...
```
Can you break these up to be one sentence per line?

stmcginnis replied:
Also, looks like you got interupted.

> and requires the user to specify...

yeazelm replied:
Whoops, I totally forgot to update this. We actually don't have a breaking change, which is great! So I'll update the docs to be accurate. I'll also look at cleaning up the markdown more thoroughly as well.

stmcginnis commented on PROVISIONING-METAL.md:
```diff
@@ -105,7 +105,30 @@ Please keep in mind that when using static addresses, DNS information must be su
+Migrating to version `3` from version `2` requires changes to how devices are defined. For example, if `[eno1]` is the current device name, changing a config from `[eno1]` to `[interface.eno1]` is all that is required. 
+
+Bonding configuration is created via a network device (`device`):
+* `bond` (map): To create a `bond` network device, specify `kind = "bond"`.
```
If you're doing an update - please add a blank line before the start of the bullet list. There is another PR that is cleaning up a lot of markdown lint errors. After that merges we could potentially add a markdownlint CI job. If we do that, it will flag this kind of thing where there is no separation before lists and code blocks.

yeazelm replied:
I'll update according the markdownlint so its consistent.

stmcginnis commented on PROVISIONING-METAL.md:
```diff
@@ -105,7 +105,30 @@ Please keep in mind that when using static addresses, DNS information must be su
+  * `device` (string for device): Defines the device the vlan should be configured on. If VLAN tagging is required, this device should recieve all IP address configuration instead of the underlying device.
+  * `id` (int): Number between 0 and 4096 specifying the vlan tag on the device
+
+
```
nit: extra blank line

Again, just a markdown lint concern, should we choose to go that route.

yeazelm replied:
Will fix.

stmcginnis commented on sources/api/netdog/src/net_config/devices/mod.rs:
```diff
@@ -0,0 +1,118 @@
+//! The devices module contains all the types of network devices that `netdog` supports. These are
```
I'm still learning, so sorry if this is a stupid question. Isn't `//!` used for inner block comments?

yeazelm replied:
I am also still learning, but from what I gathered from [this page](https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#commenting-contained-items) is that `//!` actually comments for the container, not the code that follows. This is used for module/crate level documentation. So these lines are discussing the `devices` module entirely rather than only the lines of code in this file.

stmcginnis reviewed (commented):
Looks great overall. A few minor comments. Should also clean up some of the commented out code before we merge. Nice work!

yeazelm commented on sources/api/netdog/src/net_config/devices/mod.rs:
```diff
@@ -0,0 +1,118 @@
+use serde::Deserialize;
+use vlan::NetVlanV1;
+
+// This seems to be dangling code I no longer need?
```
I intended to remove this. I will in an update.

yeazelm replied:
Removed

yeazelm commented on sources/api/netdog/src/wicked/bonding.rs:
```diff
@@ -0,0 +1,169 @@
+    t: Vec<ArpTarget>,
+}
+
+// #[derive(Debug, Clone, Serialize, PartialEq)]
```
I intended to remove this, I will pull it out in an update.

yeazelm replied:
Removed

stockholmux commented on PROVISIONING-METAL.md:
```diff
@@ -76,7 +76,7 @@ When these services fail, your machine will not connect to any cluster and will
 
 The configuration file must be valid TOML and have the filename `net.toml`.
-The first and required top level key in the file is `version`; the latest is version `2`.
+The first and required top level key in the file is `version`; the latest is version `3`.
```
Since we've created several `version`s of `net.toml` rapidly would it be helpful have a table that maps `version` with the Bottlerocket version?

I think this could be a little confusing otherwise especially if someone is only a minor version back.

yeazelm replied:
I'll add in a table for the releases that enabled this. I was able to avoid making a breaking change so we don't have any required work for users upgrading for now. Nonetheless, as we iterate this is going to be more and more important to track.

stockholmux commented on PROVISIONING-METAL.md:
```diff
@@ -105,7 +105,30 @@ Please keep in mind that when using static addresses, DNS information must be su
-Example `net.toml` with comments:
+Version `3` adds in support for bonding and vlan tagging. The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt). Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces. Physical devices like ethernet adapters are considered ...
+
+Migrating to version `3` from version `2` requires changes to how devices are defined. For example, if `[eno1]` is the current device name, changing a config from `[eno1]` to `[interface.eno1]` is all that is required. 
```
Is anything special required for version `1` to `3`?

yeazelm replied:
I edited this out, there is nothing specific a user must do to move from version `1` or `2` when moving to `3` (besides updating `version = 3`).

stockholmux reviewed (commented):
A few bits and bobs on the markdown.

yeazelm reviewed (commented):
Updated the PR from the comments so far plus tidied up some dead code.

Commit:
[provisioning-metal] Add bonding and vlan tagging doc examples

The net.toml version is bumped to 3 to accommodate the new functionality
of bonding and vlans. Version 3 does require different naming than
version 2 so version net.toml files are not drop in ready for version 3.
This also specifies which options are support for bonding and vlans.

Bonding is a complex topic and right now, we are making a deliberate
choice to support only a subset of bonding before full support is
enabled. This mostly comes in the form of only mode 1 (active-primary).

VLAN tagging is supported by creating a virtual network device on top of
another device. This allows mixing and matching of vlan tags with
whatever configuration is required. A user can choose to only configure
VLAN Tagged devices for addressing if that is required, or add in vlan
tagging on a device for special handling.

bcressey commented on sources/api/netdog/test_data/net_config/vlan/net_config.toml:
```diff
@@ -0,0 +1,37 @@
+device = "eno5"
+id = 10
+dhcp4 = true
+
```
There should be text editor settings you can change so that the file always ends up a newline, without having to add a second empty newline.

yeazelm replied:
Yeah, this one was manually done because the editor didn't do what I wanted. I can tidy it up.

etungsten commented on PROVISIONING-METAL.md:
```diff
@@ -152,13 +181,61 @@ addresses = ["192.168.14.5/24"]
+| ----------------------------- | ------------- |
+| Version 1                     | [v1.9.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.9.0)   |
+| Version 2                     | [v1.10.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.10.0) |
+| Version 3                     | v.1.12.0 |
```
```suggestion
| Version 3                     | v1.12.0 |
```
I wonder if it's best we left this as `Unreleased` since v1.12.0 doesn't exist yet and probably won't exist for a while. Then during the v1.12.0 release we'd have to remember to flip this back.

etungsten commented on sources/api/netdog/test_data/net_config/bonding/arpmon_no_targets.toml:
```diff
@@ -0,0 +1,10 @@
+[bond1.monitoring]
+arpmon-interval = 2
+arpmon-validate = "all"
+arpmon-targets = []
```
nit: missing newline unless it somehow breaks the unit tests. same for the other TOMLs/XMLs

etungsten commented on sources/api/netdog/src/wicked/mod.rs:
```diff
@@ -35,6 +47,14 @@ pub(crate) struct WickedInterface {
     #[serde(rename = "ipv6:static")]
     pub(crate) ipv6_static: Option<WickedStaticAddress>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "vlan")] // huh? not sure about this
```
```suggestion
    #[serde(rename = "vlan")]
```
Are we sure about this?

yeazelm replied:
I'll remove :)

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
-Example `net.toml` with comments:
+Version `3` adds in support for bonding and vlan tagging. 
+The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
+Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces.
```
> Version `3` also adds the concept of network devices vs network interfaces.

Is this still true in terms of config structure?

yeazelm replied:
It does, but I failed to explain what that means to the end user. `interface` really just means there is no `kind` specified. I'll add some extra detail around how one must treat interfaces differently than network devices. The other important one is you have to find the right name for an interface but can choose your own name for devices.

yeazelm replied:
I added a few more lines trying to explain that interfaces look like what came before and `virtual devices` are new and different.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
+Version `3` adds in support for bonding and vlan tagging. 
+The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
+Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces.
+Physical devices like ethernet adapters are considered "interfaces", and network devices encompass other kinds of devices like bonds and vlans.
```
This sentence can probably be omitted.

yeazelm replied:
I kind of went the other way since naming comes into play, essentially you have to give us the exact name of a physical(pre-existing) device, and you have to choose a valid name for the others. So it becomes relevant to know which one you are dealing with.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
+Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces.
+Physical devices like ethernet adapters are considered "interfaces", and network devices encompass other kinds of devices like bonds and vlans.
+The default type of device is an interface and the syntax is the same as previous versions.
+For virtual network devices, a `kind` is required.
```
Perhaps here is where we explain that if `kind` is missing, we assume the device is an interface.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
+For virtual network devices, a `kind` is required.
+Currently, `bond` and `vlan` are the two supported `kind`s.
+
+Bonding configuration is created via a network device (`device`):
```
This is no longer true. :)

yeazelm replied:
Good catch. I'll fix.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
+
+Bonding configuration is created via a network device (`device`):
+
+* `bond` (map): To create a `bond` network device, specify `kind = "bond"`.
```
This is a little bit confusing to me - `bond` isn't a top level key or map.  Perhaps this should be some sort of heading like `Bond settings`?

yeazelm replied:
Yeah, it was at one point in my head, but it no longer is. I think changing the heading is a good call.

yeazelm replied:
I updated these lines, can you confirm they make more sense now?

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
+  * `mode` (string): Currently `active-backup` is the only supported option
+  * `min-links` (int): Number of links required to bring up the device
+  * `monitoring` (map): Value must all be of `miimon` or `arpmon` type. The user must choose one type of monitoring and configure it fully in order for the bond to properly function. See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.
+    * `miimon-frequency` (int in ms): MII Monitoring frequency in miliseconds
```
As discussed offline - let's add the time scale (`milliseconds`, etc) into the setting name to make it more obvious.

yeazelm replied:
I did a bit more reading on this and realized I mis-read the ARP setting being in seconds. [From what I can tell,](https://github.com/openSUSE/wicked/blob/master/schema/bonding.xml#L44) `wicked` under the hood expects everything in milliseconds. The only thing I've found that may have given me the impression it was in seconds is from netplan's docs which call out a specific field `ArpIntervalSec` property but the default is still milliseconds. In light of everything being in milliseconds, should we just adjust everything to document all time settings are in milliseconds or still continue with `miimon-frequency-milliseconds`?

zmrow replied:
I think it probably makes sense to include the time scale in the setting name to make it obvious to the user.

yeazelm replied:
I made it `-ms` for the postfix. I think this is a reasonable compromise.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,38 @@ Please keep in mind that when using static addresses, DNS information must be su
+    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
+    * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16
+
+Vlan tagging is configured as a new network device (`device`) stacked on another device:
```
We can probably remove the `device` referenced here right?

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -152,13 +181,61 @@ addresses = ["192.168.14.5/24"]
+miimon-downdelay = 200 # 200 ms
+
+# VLAN42 is the name of the device
+[devices.VLAN42]
```
Let's remove the `devices`

zmrow commented on sources/api/netdog/src/net_config/v2.rs:
```diff
@@ -69,6 +56,8 @@ impl Interfaces for NetConfigV2 {
 
         wicked_interfaces
     }
+
+    // fn validate()
```
These two lines can be removed.

yeazelm replied:
Will remove.

zmrow commented on sources/api/netdog/src/net_config/v3.rs:
```diff
@@ -0,0 +1,106 @@
+        for (name, config) in &self.net_devices {
+            let interface = WickedInterface::from((name, config));
+
+            // if config is a Bond, for each interface: WickedInterface::new(name), add that to
+            // wicked_interfaces
```
Could we add to the comment and explain why this is necessary?

yeazelm replied:
Fixed.

zmrow commented on sources/api/netdog/src/net_config/devices/mod.rs:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+pub(crate) fn validate_addressing<D>(device: D) -> Result<()>
```
Big nit: I might move this function below all of the traits it's making use of.

yeazelm replied:
Will do.

zmrow commented on sources/api/netdog/src/net_config/devices/mod.rs:
```diff
@@ -0,0 +1,101 @@
+    Bond,
+    #[serde(rename = "vlan")]
+    Vlan,
+}
```
Nit: I think I'd move this enum up above these trait definitions and functions; it feels like it's dangling down here. :)

zmrow commented on sources/api/netdog/src/net_config/devices/interface.rs:
```diff
@@ -0,0 +1,56 @@
+impl Validate for NetInterfaceV2 {
+    fn validate(&self) -> Result<()> {
+        validate_addressing(self)?;
+        Ok(())
```
This `Ok(())` isn't necessary, the body of the function can be:
```
validate_addressing(self)
```

yeazelm replied:
Will fix.

zmrow commented on sources/api/netdog/src/net_config/static_address.rs:
```diff
@@ -11,6 +13,19 @@ pub(crate) struct StaticConfigV1 {
     pub(crate) addresses: BTreeSet<IpNet>,
 }
 
+impl Validate for StaticConfigV1 {
```
Nit: I might move this trait implementation down below the structs next to the other trait implementation (`impl TryFrom...`)

yeazelm replied:
Will move.

zmrow commented on sources/api/netdog/src/net_config/static_address.rs:
```diff
@@ -11,6 +13,19 @@ pub(crate) struct StaticConfigV1 {
+            self.addresses.iter().all(|a| matches!(a, IpNet::V4(_)))
+                || self.addresses.iter().all(|a| matches!(a, IpNet::V6(_))),
+            InvalidNetConfigSnafu {
+                reason: "'static4' may only contain IPv4 addresses"
```
This error message isn't quite right since we're making sure we have all ipv4 _or_ ipv6 addresses.

yeazelm replied:
Fixed the message

zmrow commented on sources/api/netdog/src/net_config/devices/vlan.rs:
```diff
@@ -0,0 +1,87 @@
+        // Validate its a valid vlan id - 0-4095
+        ensure!(
+            self.id <= 4095,
+            error::InvalidNetConfigSnafu {
```
Let's give a bit more context in the error message, something like `invalid vlan ID specified, must be between 0-4095`

yeazelm replied:
Fixed.

zmrow commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+        match value.to_lowercase().as_str() {
+            "bond" => Ok(BondKind(value)),
+            _ => error::InvalidNetConfigSnafu {
+                reason: "did not specify kind bond",
```
Let's give a bit more context in this error message, something like `mismatched kind: specified kind "bond" in a non-bond config`. (I'm not sold on that wording :) )

yeazelm replied:
I'm unsure anyone will ever see these messages. Instead, serde just returns that no enum matched. I'm happy to improve this message, but I would like folks to see it too. Do we have a way to bubble these errors up in aggregate on a serde failure?

zmrow commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+    fn validate(&self) -> Result<()> {
+        validate_addressing(self)?;
+
+        let interfaces_count: u32 = self.interfaces.len() as u32;
```
Is the type hint `:u32` not good enough?  You might be able to leave out the `as u32` bit.

yeazelm replied:
The type hint is not good enough, it needs the `as u32` although I can drop the type hint instead. I'm not sure if that is preferred.

zmrow commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+                ensure!(
+                    targets_length > 0 && targets_length <= 16,
+                    error::InvalidNetConfigSnafu {
+                        reason: "arpmon-targets must include at least one target"
```
Perhaps we should include acceptable values in the error message - let the user know that 1-16 targets must be specified

yeazelm replied:
Will improve the message.

zmrow commented on sources/api/netdog/src/net_config/mod.rs:
```diff
@@ -3,11 +3,14 @@
 mod static_address;
 mod v1;
-mod v2;
+pub(crate) mod v2;
+pub(crate) mod v3;
```
Do these need to be public?  Maybe I'm missing it somewhere but I don't think they're used anywhere else.

yeazelm replied:
I don't believe they do after the most recent refactor. I'll double check and if not then fix.

yeazelm replied:
Confirmed we don't need these to be public anymore. Will remove

zmrow commented on sources/api/netdog/src/wicked/mod.rs:
```diff
@@ -3,16 +3,28 @@
 pub(crate) use static_address::{WickedRoutes, WickedStaticAddress};
 use std::fs;
 use std::path::Path;
+pub(crate) use vlan::WickedVlanTag;
```
Does this need to be `pub(crate)`?  I don't think it's used anywhere else.

yeazelm replied:
I don't think it is anymore. I'll confirm and fix.

zmrow commented on sources/api/netdog/src/wicked/vlan.rs:
```diff
@@ -0,0 +1,16 @@
+}
+
+impl WickedVlanTag {
+    pub(crate) fn new(device: InterfaceName, id: u16) -> Option<Self> {
+        Some(WickedVlanTag { device, id })
+    }
```
It's a bit unconventional to have a constructor return an `Option` - this should probably return a `Self` and it can be wrapped with `Some()` if needed elsewhere.

yeazelm replied:
Oh, yeah, now that you say that, that makes sense. I'll move it to where it should be.

zmrow reviewed (commented):
Nice work!

zmrow commented on sources/api/netdog/src/net_config/static_address.rs:
```diff
@@ -1,6 +1,8 @@
+use super::error::{InvalidNetConfigSnafu, Result as ValidateResult};
```
Nit: Here we can use `super::{error, Result}` rather than the specific error variant and aliasing the `Result`.

yeazelm replied:
https://github.com/bottlerocket-os/bottlerocket/pull/2596/files/ae2e4222e3c8f47f5d2b664ad51271e150ee7877#diff-12e73d418c8b0b97c34a00dbe0ba9ee367e114a6baef0edd0dac915b00c68557L65 - I was having trouble since we defined our own `Result` here at the bottom of the file. I'm unsure if that is actually desired at this point.

zmrow replied:
Nm, I see what is going on here.  Disregard

zmrow commented on sources/api/netdog/src/wicked/mod.rs:
```diff
@@ -3,16 +3,28 @@
+use crate::wicked::bonding::{
+    WickedArpMonitoringConfig, WickedBondMode, WickedMiiMonitoringConfig,
+};
+pub(crate) use bonding::WickedBond;
```
Nit: I might have missed it, but does this need to be `pub`?

yeazelm replied:
It looks like this one no longer needs to be `pub`. I'll fix it.

webern commented on sources/api/netdog/src/wicked/vlan.rs:
```diff
@@ -0,0 +1,16 @@
+
+#[derive(Debug, Clone, Serialize, PartialEq)]
+pub(crate) struct WickedVlanTag {
+    #[serde(rename = "$unflatten=device")]
```
Is this a serde feature, or is this the actual name of the field? `$unflatten`

yeazelm replied:
Yes, its a feature with [quickxml and serde](https://crates.io/crates/quick-xml/0.23.0-alpha3) actually, see "Unflattening structs into verbose XML" on that page. This is required to get the xml to be how wicked expects to parse it.

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+    }
+}
+
+// Currently mode 1 (active-backup) is only supported but eventually 0-6 could be added
```
```suggestion
// Currently only mode 1 (active-backup) is supported but eventually 0-6 could be added
```

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+pub(crate) enum BondMode {
+    #[serde(rename = "active-backup")]
+    ActiveBackup,
+}
```
I always misspell kebab, I wonder if I got it right for once.

```suggestion
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub(crate) enum BondMode {
    ActiveBackup,
}
```

yeazelm replied:
I avoided this specific case for the future when we would have `802.3ad` but maybe we can specify a specific override for that one? I do like letting serde take care of all of this for us. I'll move to that for now anyway.

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+    Backup,
+    #[serde(rename = "none")]
+    None,
+}
```
(or lowercase...)

This is generally applicable anywhere you have used `[serde(rename="whatever")]` when it could actually be done with `serde(rename_all = "kebab-case")`

```suggestion
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub(crate) enum ArpValidate {
    Active,
    All,
    Backup,
    None,
}
```

yeazelm replied:
Yeah, this is much nicer, I'll change to this.

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+
+#[derive(Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct NetBondV1 {
```
What if we adopted the convention to have the "current" version of the struct un-suffixed and only add the suffix when V2 exists? Just a thought to make things a little cleaner. @zmrow ?

```suggestion
pub(crate) struct NetBond {
```

yeazelm replied:
It's pretty consistent at this point to have `V1` on the end of these types in netdog code. It would be a bit of a refactor to make that consistent going forward. If we do want to make this change, I highly recommend we do that in a follow on PR because that would increase the surface of this change well outside the original scope and this is already a pretty big PR.

zmrow replied:
I do like the explicit nature of a versioned, i.e `*V1` struct, especially in this context where there are quite a few structs to keep track of.

If there are strong feelings about this, I'd prefer to do that refactor in a separate PR as @yeazelm mentioned.  There's already _alot_ of change here.

yeazelm replied:
I think I'll punt this for a topic we can sort out in a later PR but not this one.

bcressey commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,43 @@ Please keep in mind that when using static addresses, DNS information must be su
+    * `miimon-downdelay` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
+    * `arpmon-interval` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0
+    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
+    * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16
```
nit: maybe indent the miimon and arpmon options another level

yeazelm replied:
They are indented one, did you want more or for them to be inline with the rest of the options?

bcressey replied:
Not sure if it wasn't rendering right for me earlier, but you're right and it looks fine to me now.

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+    pub(crate) interfaces: Vec<InterfaceName>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(try_from = "String")]
+pub(crate) struct BondKind(String);
+
+impl TryFrom<String> for BondKind {
+    type Error = error::Error;
+
+    fn try_from(value: String) -> Result<Self> {
+        match value.to_lowercase().as_str() {
+            "bond" => Ok(BondKind(value)),
+            _ => error::InvalidNetConfigSnafu {
+                reason: format!("kind of {} does not match bond", value.as_str()),
+            }
+            .fail(),
+        }
+    }
+}
```
Do we really want to treat `Bond` and `boND` as valid? It's not how we do it with other enumerable strings, e.g. in the API, where we require the casing to be correct. Also then we are storing the funkily-cased string for later use.

If `kind` is dead code anyway (only used for validation during deserialization) then presumably there is no reason to store the user's string. ~~Also, if we wanted we could probably do code re-use and have~~

```suggestion
#[derive(Debug, Deserialize)]
#[serde(try_from = "String")]
pub(crate) struct BondKind;

impl TryFrom<String> for BondKind {
    type Error = error::Error;

    fn try_from(value: String) -> Result<Self> {
        match value.as_str() {
            "bond" => Ok(BondKind),
            _ => error::InvalidNetConfigSnafu {
                reason: format!("kind of '{}' does not match 'bond'", value.as_str()),
            }
            .fail(),
        }
    }
}
```

Also, I'm sad we can't do this... it's probably coming to the language at some point:

```rust
#[derive(Debug, Deserialize)]
#[serde(try_from = "String")]
pub(crate) struct ValidString<const S: &'static str>;

impl<S> TryFrom<String> for ValidString<const S: &'static str>
{
    type Error = error::Error;

    fn try_from(value: String) -> Result<Self> {
        match value.as_str() {
            S => Ok(BondKind),
            _ => error::InvalidNetConfigSnafu {
                reason: format!("kind of '{}' does not match '{}'", value, S),
            }
            .fail(),
        }
    }
}
```

yeazelm replied:
For the naming being case sensitive. This is a key we made up (`kind`) only has a helper to both the user and us to make sure we deserialize the intended type of device. Not accepting "Bond" or "BOND" feels slightly more strict than needed. It can be infuriating to lose time to something as silly as a capitalization difference when configuring your system, and since a user has to try a bare metal boot cycle which can take minutes to try things, I have aimed to be as forgiving as possible. I can be talked into keeping the case strict to stay consistent with everything else in the api server, but I really think its a bad customer experience. 

On the storing the string side, I think you are right, I had assumed that `try_from` required you to actually be doing this on a resulting string, but it appears that is not necessary. This is still not ideal, and @zmrow and I have some ideas on cleaning this up further, so I may just take this refactor or do more depending on how much we like this slightly better improvement.

yeazelm replied:
I ended up going a completely different way in the logic here. Not sure if its better but would like your take. It feels both better and worse to just implement this logic in the deserialize logic directly rather than use a separate struct and try_from for this.

yeazelm replied:
I was able to clean this up even more with @zmrow so I think this turned out nicely now!

webern reviewed (commented):
I need to save a few thoughts and move on to other things. I'll continue with this PR when I can. Nice work!

bcressey commented on sources/api/netdog/src/wicked/mod.rs:
```diff
@@ -90,6 +113,122 @@ impl WickedInterface {
+
+impl From<(&InterfaceName, &NetBondV1)> for WickedInterface {
+    fn from(device_tup: (&InterfaceName, &NetBondV1)) -> Self {
+        // Starts the same as interface, we might find a way to refactor this
+        let config = &device_tup.1;
+        let mut wicked_interface = WickedInterface::new(device_tup.0.clone());
+        wicked_interface.ipv4_dhcp = config.dhcp4.clone().map(WickedDhcp4::from);
+        wicked_interface.ipv6_dhcp = config.dhcp6.clone().map(WickedDhcp6::from);
+
+        // Based on the existence of static addresses and routes, create the ipv4/6_static
+        // struct members.  They must be `Option`s because we want to avoid serializing empty
+        // tags into the config file
+        let maybe_routes = config.routes.clone().map(WickedRoutes::from);
+        let maybe_ipv4_static = WickedStaticAddress::maybe_new(
+            config.static4.clone(),
+            maybe_routes.as_ref().and_then(|s| s.ipv4.clone()),
+        );
+        let maybe_ipv6_static = WickedStaticAddress::maybe_new(
+            config.static6.clone(),
+            maybe_routes.as_ref().and_then(|s| s.ipv6.clone()),
+        );
+        wicked_interface.ipv4_static = maybe_ipv4_static;
+        wicked_interface.ipv6_static = maybe_ipv6_static;
```
We've got three copies of this code now, it'd be good to refactor even if it's just with a macro.

(make that four - there's still a copy in `net_config/v2.rs`)

yeazelm replied:
I can look into making it a macro.

bcressey commented on sources/api/netdog/src/net_config/v3.rs:
```diff
@@ -0,0 +1,106 @@
+        for (name, config) in &self.net_devices {
+            let interface = WickedInterface::from((name, config));
+
+            // If config is a Bond, we will generate the interface configuration. For each interface:
+            // WickedInterface::new(name), add that to wicked_interfaces
+            if let NetworkDeviceV1::BondDevice(b) = config {
+                for device in &b.interfaces {
+                    let mut wicked_sub_interface = WickedInterface::new(device.clone());
+                    wicked_sub_interface.link = Some(WickedLinkConfig {
+                        master: name.clone(),
+                    });
+
+                    wicked_interfaces.push(wicked_sub_interface)
+                }
+            }
```
Not sure I follow this comment or the logic - if a "bond" interface declares `eno11` and `eno12` as its interfaces, we are generating a minimal wicked config for it?

yeazelm replied:
This is correct, I should probably word this better. As it stands, we generate the entire xml structure for each interface within a bond because wicked is quite specific about what that looks like and we have all the information to do it on behalf of the user. But this is the downside of doing this generation of xml for the user: what to do with configuration the user has provided us. We can attempt to use what they gave us, throw it away and warn them, try to merge them somehow and hope it's valid, or refuse to proceed and make them resolve the conflicting configuration. I opted for the one that forces the user to resolve this, rather than be surprised by the decision we make.

bcressey commented on sources/api/netdog/src/net_config/v3.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+}
+
+#[allow(clippy::to_string_in_format_args)]
+impl Validate for NetConfigV3 {
```
It would be good to switch `NetConfigV2` and `NetConfigV1` over to the new validation implementation, so we don't have redundant copies of the logic that's now in `validate_addressing`.

yeazelm replied:
I went back and forth on whether to do this. I was attempting to avoid touching more code than strictly necessary to achieve this new functionality but I can refactor this out. I'll add the changes as a separate commit for it though because its now getting into refactoring and cleaning up the code base vs simply adding in the needed code for bonding and vlans.

yeazelm replied:
I've moved the v2 validation to the new style but v1 didn't have static addressing which is where most of this actually came in. So it is unchanged.

bcressey commented on sources/api/netdog/src/net_config/v3.rs:
```diff
@@ -0,0 +1,106 @@
+            // Bonds create the interfaces automatically, specifying those interfaces would cause a
+            // collision so this emits an error
+            if let NetworkDeviceV1::BondDevice(config) = device {
+                for interface in &config.interfaces {
+                    if self.net_devices.contains_key(interface) {
+                        return error::InvalidNetConfigSnafu {
+                            reason: format!(
+                                "{} in bond {} cannot be manually configured",
+                                interface.to_string(),
+                                name.to_string()
+                            ),
+                        }
+                        .fail();
```
Is it legal for one of the bond interfaces to appear as the "device" field in a VLAN device?

yeazelm replied:
I don't believe so. But this logic wouldn't catch that, it only catches if a user specified the raw physical interface directly. If they specify 
```toml
[bond0]
kind = bond,
interfaces = ["eno1", "eno2"]
....

[tag1]
kind = vlan
device = "eno1"
....
```
We wouldn't catch them both pointing at it. We'd only catch if they specified `eno1` directly like so:
```toml
[eno1]
dhcp4 = true
```
Its a little more tricky to do this deep validation but I can add it (after I confirm that truly won't work).

yeazelm replied:
This turns out to be a tricky edge case to test for, in order to successfully test a multi-device configuration, we need to do this carefully. I can take this as something to block the PR if you think its important, otherwise I'm happy to improve our error handling in a follow on PR.

yeazelm replied:
After I wrote this, it occurred to me that I could create a `HashSet` of all the names and simply check for them there. It forces another loop over all the devices but from here we could do other checks if we need. I'm open to improvements since we are now looping twice but I couldn't seem to find a good way to ensure bonds don't refer to devices in vlans if we haven't found the conflicting vlan yet in the loop. In this case, we don't expect many interfaces to be configured so looping twice shouldn't be too high of a penalty to catch this edge case.

bcressey commented on sources/api/netdog/src/net_config/devices/vlan.rs:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+impl HasStatic for &NetVlanV1 {
+    fn has_static(&self) -> bool {
+        self.static4.is_some() || self.static6.is_some()
+    }
+
+    fn validate_static4(&self) -> Result<()> {
+        if let Some(config) = &self.static4 {
+            config.validate()?
+        }
+        Ok(())
+    }
+
+    fn validate_static6(&self) -> Result<()> {
+        if let Some(config) = &self.static6 {
+            config.validate()?
+        }
+        Ok(())
+    }
+}
+
+impl HasDhcp for &NetVlanV1 {
+    fn has_dhcp(&self) -> bool {
+        self.dhcp4.is_some() || self.dhcp6.is_some()
+    }
+}
+
+impl HasRoutes for &NetVlanV1 {
+    fn has_routes(&self) -> bool {
+        self.routes.is_some()
+    }
+}
```
For me, all these helper traits are hurting the clarity of the code.

I'd recommend looking at a derive macro that can be reused across the types, so you can just write:
```
#[derive(Debug, Deserialize, Validate)]
pub(crate) struct NetVlanV1 {
}
```

Then reviewers can focus on two things:
* do all the types that need to derive `Validate`?
* does the single implementation of `Validate` cover all the combinations?

Right now, it's more like this:
* is `HasRoutes` implemented the same way for `NetInterfaceV2`, `NetVlanV1`, `NetBondV1`?
* is `HasStatic` ... ?
* is `HasDhcp` ... ?
* do all the structs that implement all three also call `validate_addressing`?

yeazelm replied:
That's fair, I'll look at implementing a derive macro.

yeazelm replied:
I didn't go as far as a derive, but I did clean up the code and move most of this boiler plate into a single trait that is generated for those devices that need it. So now someone should only need to reason that this has been added to the right places and the macro correctly generates the logic required to validate IP Addressing.

bcressey commented on sources/api/netdog/src/net_config/devices/vlan.rs:
```diff
@@ -0,0 +1,87 @@
+    #[allow(dead_code)]
+    kind: VlanKind,
+    pub(crate) device: InterfaceName,
+    pub(crate) id: u16,
```
potentially this could be validated during deserialization, like "kind" is

yeazelm replied:
I'm still not certain where the optimal place is to enforce these constraints, especially when typos or naive misconfigurations could happen. There are some other refactors going around the "kind" validation as well. I'll see if it makes sense to shift it there.

yeazelm replied:
I moved it into the custom deserializer, thanks for the tip!

bcressey commented on sources/api/netdog/src/wicked/bonding.rs:
```diff
@@ -0,0 +1,163 @@
+        };
+        Self {
+            mode,
+            devices: slaves,
```
```suggestion
        let s = SubDevices {
            devices: sub_devices,
        };
        Self {
            mode,
            devices: s,
```

bcressey commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,184 @@
+                    }
+                )
+            }
+            BondMonitoringConfig::ArpMon(config) => {
+                ensure!(
+                    config.interval > 0,
+                    error::InvalidNetConfigSnafu {
+                        reason: "arpmon-interval of 0 disables Arp Monitoring, either set a value or configure Mii Monitoring"
+                    }
+                );
+                // If using Arp Monitoring, 1-16 targets must be specified
+                let targets_length: u32 = config.targets.len() as u32;
+                ensure!(
+                    targets_length > 0 && targets_length <= 16,
+                    error::InvalidNetConfigSnafu {
+                        reason: "arpmon-targets must include between 1 and 16 targets"
+                    }
+                )
+            }
```
Since these validations don't depend on any fields from the parent struct, I'd prefer for them to have their own validate methods that get called here.

yeazelm replied:
I can move them to their own methods for those structs.

bcressey commented on sources/api/netdog/src/net_config/devices/vlan.rs:
```diff
@@ -0,0 +1,87 @@
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(try_from = "String")]
+pub(crate) struct VlanKind(String);
```
Can you just use an empty newtype here, and discard the string?
```suggestion
#[serde(try_from = "String")]
pub(crate) struct VlanKind();
```

(same with `BondKind`)

yeazelm replied:
I had assumed that we needed to store the string to use `try_from` on String but it looks like you don't. This seems to be what @webern was referring to as well. I'll work on this a bit to refine.

yeazelm commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,221 @@
+    pub(crate) interfaces: Vec<InterfaceName>,
+}
+
+impl<'de> Deserialize<'de> for NetBondV1 {
```
I'm not 100% convinced this is better than the TryFrom String bits, but this allows us to catch the specifics in line during deserialization. We might be able to improve the casting one struct as another too at the end of this function but wanted to get it out to review.

yeazelm replied:
@zmrow  helped me on the piece I was missing, this is now much cleaner and I'm happy with where it ended up.

webern commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,43 @@ Please keep in mind that when using static addresses, DNS information must be su
-Example `net.toml` with comments:
+Version `3` adds in support for bonding and vlan tagging. 
+The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
+Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces.
```
nit: style consistency

```suggestion
Future support may include other bonding options - pull requests are welcome!
Version `3` also adds the concept of network devices vs network interfaces.
```

yeazelm replied:
I'll fix it.

webern commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,43 @@ Please keep in mind that when using static addresses, DNS information must be su
+  * `interfaces` (list of quoted strings of interfaces): Which interfaces should be added to the bond (i.e. `["eno1"]`). The first in the list is considered the default `primary`.
+  * `mode` (string): Currently `active-backup` is the only supported option
+  * `min-links` (int): Number of links required to bring up the device
+  * `monitoring` (map): Value must all be of `miimon` or `arpmon` type. The user must choose one type of monitoring and configure it fully in order for the bond to properly function. See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.
```
The main nit is grammar (Values), but also the long line could be broken up.

```suggestion
  * `monitoring` (map): Values must all be of `miimon` or `arpmon` type.
    The user must choose one type of monitoring and configure it fully in order for the bond to properly function.
    See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.
```

yeazelm replied:
I'll fix this as well.

webern commented on sources/api/netdog/test_data/net_config/bonding/net_config.toml:
```diff
@@ -0,0 +1,35 @@
+[bond1.monitoring]
+arpmon-interval = 200
+arpmon-validate = "all"
+arpmon-targets = ["192.168.1.1", "10.0.0.2"]
```
Are these always ipv4? Should you have one ipv6 value in the test data here?

yeazelm replied:
Yeah I believe they could be IPv6 as well, but I just never saw any examples. I can probably confirm that is valid but the kernel docs only use IPv4 (which might just date when the doc was written).

webern commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,43 @@ Please keep in mind that when using static addresses, DNS information must be su
+    * `miimon-updelay-ms` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
+    * `miimon-downdelay-ms` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
+    * `arpmon-interval-ms` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0
+    * `arpmon-validate-ms` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
```
Is this a milliseconds value?

```suggestion
    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
```

yeazelm replied:
I was too aggressive in my adding `-ms`. I'll pull them out, you are correct, the validate is not an `-ms`.

webern commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,43 @@ Please keep in mind that when using static addresses, DNS information must be su
+    * `miimon-downdelay-ms` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
+    * `arpmon-interval-ms` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0
+    * `arpmon-validate-ms` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
+    * `arpmon-targets-ms` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16
```
Same here, am I not understanding what the `ms` suffix means? I thought it was milliseconds.

```suggestion
    * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16
```

yeazelm replied:
Same as above, was too aggressive.

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,161 @@
+
+// Currently only mode 1 (active-backup) is supported but eventually 0-6 could be added
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub(crate) enum BondMode {
+    ActiveBackup,
+}
```
At the moment I'm not sure if you're using this as a number elsewhere, but if you are, this is a way that you can do it.

```suggestion
#[serde(rename_all = "kebab-case")]
#[repr(u8)]
pub(crate) enum BondMode {
    ActiveBackup = 1,
}

impl BondMode {
    pub fn as_u8(&self) -> u8 {
        self as u8
    }
}
```

yeazelm replied:
We are actually using the String right now. But we could swap to number, I believe the two are interchangeable in the wicked xml but I didn't bother checking since the strings worked. This is nice though, and if we wanted to support folks choosing either, this might also work.

webern commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,161 @@
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
```
There are probably a few enums like this in the PR. It can be handy when writing code if "simple" enums implement `Copy` since they are nothing but an integer under the hood.

```suggestion
#[derive(Clone, Copy, Debug, Deserialize)]
```

yeazelm replied:
Ok, I'll take a look at adding them to the Enums that are "simple" like you call out.

webern commented on sources/api/netdog/src/net_config/static_address.rs:
```diff
@@ -48,6 +50,19 @@ impl TryFrom<String> for RouteTo {
+            self.addresses.iter().all(|a| matches!(a, IpNet::V4(_)))
+                || self.addresses.iter().all(|a| matches!(a, IpNet::V6(_))),
+            InvalidNetConfigSnafu {
+                reason: "static configuration must only contain all IPv4 or all IPv6 addresses"
```
I think this answers a question I asked elsewhere.

In general I'm sad to see all of this validation code. And I'm sure I was asked for help with this and probably didn't come up with anything, but there's always a way to avoid "validate after construction", especially with Rust. Just something to keep in mind in future designs. 

(Also, I realize that making it compatible with the serialization shape is challenging)

Instead of `Vec<V4orV6Enum>`...


```
enum IpV4OrIpV6List {
    IpV4(Vec<IpV4Struct>),
    IpV6(Vec<IpV6Struct>),
}
```

Which is one example of eliminating a `validate()` call.

webern commented on sources/api/netdog/src/net_config/test_macros/bonding.rs:
```diff
@@ -0,0 +1,81 @@
+
+            gen_boilerplate!($version, "bonding");
+
+            #[test]
+            fn ok_config() {
```
Generating test code like this confounds IDE tooling. Just something to keep in mind in the future. One cannot step through code or click on the green "run this test" button with generated code.

That being said, I love the fact that so much testing is taking place and the test configuration files serve as excellent documentation.

yeazelm replied:
Yeah, CLion does do some interesting things where you can view the generated file on some failures, I'm not sure how I got it doing it, but it did help. It's not the most ergonomic experience when dealing with these tests but I also don't think I would rather have the alternative. And its great to have tests that can map across versions pretty easily as we move forward.

webern commented on sources/api/netdog/src/net_config/v3.rs:
```diff
@@ -0,0 +1,119 @@
+            // collision so this emits an error for any that are found
+            if let NetworkDeviceV1::BondDevice(config) = device {
+                for interface in &config.interfaces {
+                    if !interface_names.insert(interface) {
```
Is this a HashSet? If so there is a `contains` function. Maybe I misunderstand what is happening here

https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.contains

yeazelm replied:
In this case I'm actually checking if the [insert found one already all in the set](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert), if it already exists, then its a failure, but we want the set to grow as we find interfaces (two bonds both wanting the same interface is a primary concern of this function). Your comment shows I may be trying to be too clever without at least providing a clue in a comment that we are checking the return of this `insert`.

webern commented on sources/api/netdog/src/net_config/v3.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+}
+
+#[allow(clippy::to_string_in_format_args)]
```
When you delete the `.to_string()` calls do you get some other clippy message? If so that must be a false positive (assuming the types implement `Display`).

yeazelm replied:
Yeah, it complains differently about this:
```rust
error[E0277]: `interface_name::InterfaceName` doesn't implement `std::fmt::Display`
  --> api/netdog/src/net_config/v3.rs:81:33
   |
81 | ...                   interface,
   |                       ^^^^^^^^^ `interface_name::InterfaceName` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `interface_name::InterfaceName`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: this error originates in the macro `$crate::__export::format_args` which comes from the expansion of the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)
```

webern replied:
Ah, the "right" thing to do would be to implement `Display` for `interface_name::InterfaceName` instead of implementing a `to_string` function (if that's what you did)

bcressey commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,46 @@ Please keep in mind that when using static addresses, DNS information must be su
+    * `miimon-frequency-ms` (int in ms): MII Monitoring frequency in milliseconds
+    * `miimon-updelay-ms` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
+    * `miimon-downdelay-ms` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
+    * `arpmon-interval-ms` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0
```
"(int in ms)" is hard for me to parse, like it implies there's some sort of numeric conversion that has to happen.

```suggestion
    * `miimon-frequency-ms` (integer): MII Monitoring frequency in milliseconds
    * `miimon-updelay-ms` (integer): MII Monitoring delay before the link is enabled after link is detected in milliseconds
    * `miimon-downdelay-ms` (integer): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
    * `arpmon-interval-ms` (integer): Number of milliseconds between intervals to determine link status, must be greater than 0
```

bcressey commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,46 @@ Please keep in mind that when using static addresses, DNS information must be su
+    * `miimon-updelay-ms` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
+    * `miimon-downdelay-ms` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
+    * `arpmon-interval-ms` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0
+    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
```
nit: since this is a TOML file, it might be better to only specify types and values that TOML understands

```suggestion
    * `arpmon-validate` (one of `"all"`, `"none"`, `"active"`, `"backup"`): What packets should be used to validate link
```

bcressey commented on PROVISIONING-METAL.md:
```diff
@@ -152,13 +189,65 @@ addresses = ["192.168.14.5/24"]
+|-------------------------------|---------------------------------------------------------------------------------|
+| Version 1                     | [v1.9.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.9.0)   |
+| Version 2                     | [v1.10.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.10.0) |
+| Version 3                     | Unreleased                                                                      |
```
nit: it might be better to go ahead and optimistically set this to `1.12.0`, and just fix it up prior to merge if it doesn't make the release

Otherwise we might miss changing it in a follow-up PR prior to cutting the release.

yeazelm replied:
I originally had it that way. We had expected to merge this a while ago and have other edits in the repo since then and flip then. Since we are still in PR, it makes sense to just change it for the next release.

yeazelm replied:
I decided to leave this for now, I'll be tracking quite a bit for the next release and plan to make sure this makes it into the release changes for whatever release it does make.

bcressey commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,161 @@
+    fn validate(&self) -> Result<()> {
+        validate_addressing(self)?;
+
+        let interfaces_count: u32 = self.interfaces.len() as u32;
```
You are doing a cast from the `usize` returned by `Vec`'s `len` method, to `u32` which is defined but can result in truncation. It might be better to rework the code a bit so you can keep `interfaces_count` as a `usize` - possibly by tweaking the comparsion to `min_links` if necessary.

yeazelm replied:
Good catch, I'll move them to `usize`

bcressey commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,161 @@
+        validate_addressing(self)?;
+
+        let interfaces_count: u32 = self.interfaces.len() as u32;
+        ensure!(
+            interfaces_count > 0,
+            error::InvalidNetConfigSnafu {
+                reason: "bonds must have 1 or more interfaces specified"
+            }
+        );
+        if let Some(min_links) = self.min_links {
+            ensure!(
+                min_links <= interfaces_count,
+                error::InvalidNetConfigSnafu {
+                    reason: "min-links is greater than number of interfaces configured"
+                }
+            )
+        }
```
Can these not also be checked at deserialization time, the way "kind" is?

yeazelm replied:
These can be moved into custom deserializers. I have avoided moving everything in to custom derserializers due to [this serde bug](https://github.com/serde-rs/serde/issues/2157). Until we resolve this, deserialization errors are masked and therefore it's actually providing less feedback to the user. I would like to leave them for now with the understanding we will move these when we resolve the error messaging issues. This will at least result in a user seeing a specific validation failure as it is written.

zmrow replied:
^ This sounds like the right strategy.

yeazelm replied:
https://github.com/bottlerocket-os/bottlerocket/issues/2657 I cut this issue to cover fixing this long term. I think this improves many things, not just this particular case.

bcressey commented on sources/api/netdog/src/net_config/devices/bonding.rs:
```diff
@@ -0,0 +1,161 @@
+                }
+            )
+        }
+        // Validate monitoring configuration
+        match &self.monitoring_config {
+            BondMonitoringConfig::MiiMon(config) => config.validate()?,
+            BondMonitoringConfig::ArpMon(config) => config.validate()?,
+        }
```
Same question here - can we move these checks into the deserializers for `MiiMonitoringConfig` and `ArpMonitoringConfig`?

It seems like this should be possible in other places (e.g. `NetInterfaceV2`) as long as only the one struct and its fields are involved in validation.

yeazelm replied:
As I mentioned above, I do think we can and should do this once we fix our issues with the enum failure messaging.

yeazelm replied:
https://github.com/bottlerocket-os/bottlerocket/issues/2657 Covers this work.

bcressey commented on sources/api/netdog/src/net_config/devices/mod.rs:
```diff
@@ -0,0 +1,128 @@
+    Ok(())
+}
+
+// For all devices that have IP Addressing available, generate the trait implementation
+macro_rules! generate_addressing_validation {
+    ($name:ty) => {
```
The macro and trait look OK to me. However, the generated method could potentially also be invoked in the custom deserializer, which when combined with everything else could eliminate the need for the `Validate` trait and invocation.

zmrow replied:
Yeah - that's probably a good move!  Eliminates some boilerplate as well.

yeazelm replied:
This makes sense to me. We can do the clean up on this macro call with the rest of the move to Deserialize and that should remove the need for validate all in one change.

yeazelm replied:
https://github.com/bottlerocket-os/bottlerocket/issues/2657 covers this work.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,46 @@ Please keep in mind that when using static addresses, DNS information must be su
+If no `kind` is specified, it is assumed to be an interface.
+Currently, `bond` and `vlan` are the two supported `kind`s.
+Virtual network devices are created, and therefore a name has to be chosen.
+Names for virtual network devices must conform to [kernel naming requirements](https://elixir.bootlin.com/linux/v5.10.102/source/include/uapi/linux/if.h#L33).
```
This link isn't particularly helpful - perhaps we should enumerate the short list of requirements here?

yeazelm replied:
I'll add it to be more explicit instead.

zmrow commented on PROVISIONING-METAL.md:
```diff
@@ -106,10 +107,46 @@ Please keep in mind that when using static addresses, DNS information must be su
+Version `3` adds in support for bonding and vlan tagging. 
+The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
+Future support may include other bonding options - pull requests are welcome! 
+Version `3` also adds the concept of network devices vs network interfaces.
```
Here we say `network devices vs. network interfaces`, but the following sentence refers to an interface as a type of device.

Perhaps we should say something like `Version 3 adds the concept of virtual network devices in addition to interfaces`

zmrow reviewed (approved):
Nice work!

Let's open an issue re:  [better error messaging when deserializing untagged enums](https://github.com/serde-rs/serde/issues/2157).  Pulling that in will allow us to move all validation into `Deserialize` implementations and remove the `Validate` trait since our error messages won't be masked by the top-level enum deserialize.

yeazelm commented:
I added https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/419 to track the work to fix the error messaging for enums so we can move all the rest of the validation into the Deserialization.
